hoot-core/src/main/cpp/hoot/core/algorithms/WayMatchStringMerger.cpp

```diff
@@ -26,12 +26,389 @@
+    {
+      throw InternalErrorException(QString("Not a way match %1, %2").arg(i).arg(
+        hoot::toString(ws1->at(i).getStart().getWay()->getElementId())));
+    }
+    // if we found the end of the match
+    if (ws1->at(i).getEnd().getWay() == split2.getWay())
+    {
+      lastSm->end = split2;
+      _addSublineMapping(lastSm);
+      foundEnd = true;
+    }
+    // if we aren't at the end yet
+    else
+    {
+      // add the rest of the subline to the mapping
+      lastSm->end = ws1->at(i).getEnd();
+      _addSublineMapping(lastSm);
+    }
+    lastSm.reset();
+  }
+}
+
+WayLocation WayMatchStringMerger::_findNodeLocation2(WayStringPtr ws, ElementId nodeId)
+{
+  assert(nodeId.getType() == ElementType::Node);
+
+  // if we haven't created the index, create it.
+  if (_nodeToWayLocation2.size() == 0)
+  {
+    // go through the way string and look at each subline.
+    for (int i = 0; i < ws->getSize(); ++i)
+    {
+      WaySubline subline = ws->at(i);
+      ConstWayPtr w = subline.getWay();
+      // go through all the segments that are part of this subline
+      for (int j = subline.getFormer().getSegmentIndex();
+        j <= subline.getLatter().getSegmentIndex(); ++j)
+      {
+        WayLocation wl(_map, w, j, 0.0);
+        if (subline.contains(wl))
+        {
+          // only record the first instance.
+          if (_nodeToWayLocation2.contains(ElementId::node(w->getNodeId(j))) == false)
+          {
+            assert(wl.isNode() && wl.getNode()->getId() == w->getNodeId(j));
+            _nodeToWayLocation2[ElementId::node(w->getNodeId(j))] = wl;
+          }
+        }
+      }
+    }
+  }
+
+  if (_nodeToWayLocation2.contains(nodeId) == false)
+  {
+    throw IllegalArgumentException("Unable to find node in way string 2.");
+  }
+
+  return _nodeToWayLocation2[nodeId];
+}
+
+void WayMatchStringMerger::mergeNode(ElementId scrapNode)
+{
+  LOG_VAR(scrapNode);
+  LOG_VAR(_map->getNode(scrapNode));
+  // find the first instance of scrapNode in way 2
+  WayLocation wl2 = _findNodeLocation2(_mapping->getWayString2(), scrapNode);
+
+  // map the WayLocation of scrapNode to way string 1
+  WayLocation wl1 = _mapping->map2To1(wl2);
+
+  // move scrapNode to the WayLocation in way string 1
+  _moveNode(scrapNode, wl1);
+
+  // notify _sublineMapping that way string 1 changed.
+  _rebuildWayString1();
+}
+
+void WayMatchStringMerger::mergeTags()
+{
+  if (!_tagMerger)
+  {
+    throw IllegalArgumentException("You must specify the tag merger first.");
+  }
+
+  // go through all the way mappings
+  for (QMultiMap<WayPtr, SublineMappingPtr>::iterator it = _sublineMapping.begin();
+    it != _sublineMapping.end(); ++it)
+  {
+    LOG_VAR(it.value());
+    // merge the tags (order matters)
+    Tags mergedTags = _tagMerger->mergeTags(it.value()->newWay1->getTags(),
+      it.value()->way2->getTags(), it.value()->newWay1->getElementType());
+    // set the new tags.
+    it.value()->newWay1->setTags(mergedTags);
+  }
+}
```

Comment:
WayMatchStringMerger::setTagMerger sets the tag merger that will be used. It must be called before mergeTags() is called.